lib/bolt/applicator.rb

```diff
@@ -66,6 +66,54 @@ def query_resources_task
       end
     end
 
+    def fork_compile(catalog_input)
+      out_r, out_w = IO.pipe
+      err_r, err_w = IO.pipe
+
+      pid = fork do
+        begin
+          out_r.close
+          err_r.close
+          $stdout = out_w
+          $stderr = err_w
+
+          # This should be preloaded in most cases.
+          require 'bolt/catalog'
+
+          catalog_input = JSON.parse(catalog_input.to_json)
+          catalog = Bolt::Catalog.new.from_request(catalog_input)
+          # This seems to be a string in ruby 2.3
+          if catalog.is_a?(String)
+            catalog = JSON.parse(catalog)
+          end
+          puts JSON.pretty_generate(catalog)
+        # forks don't exit non-zero on unhandled exceptions
+        rescue StandardError => e
+          $stderr.puts "#{e.class}: #{e.message}"
+          $stderr.puts e.backtrace.join("\n")
+          $stderr.flush
+          $stdout.flush
+          Kernel.exit!(1)
+        end
+      end
+
+      _, status = Process.wait2(pid)
+
+      out_w.close
+      err_w.close
+
+      return [out_r.read, err_r.read, status]
```

Comment:
This can deadlock if either `stdout` or `stderr` fills its buffer.

Reply:
I think the puppet solution nick points to above may be better. I think `Thread.new { out.read }` may not exit properly in out doesn't get an eof.